packages/deployments/contracts/contracts/test/OZERC20.sol

```diff
@@ -287,4 +318,67 @@ contract ERC20 is IERC20 {
     address _to,
     uint256 _amount
   ) internal virtual {}
+
+  /**
+   * @dev See {IERC20Permit-permit}.
+   * @notice Sets approval from owner to spender to value
+   * as long as deadline has not passed
+   * by submitting a valid signature from owner
+   * Uses EIP 712 structured data hashing & signing
+   * https://eips.ethereum.org/EIPS/eip-712
+   * @param _owner The account setting approval & signing the message
+   * @param _spender The account receiving approval to spend owner's tokens
+   * @param _value The amount to set approval for
+   * @param _deadline The timestamp before which the signature must be submitted
+   * @param _v ECDSA signature v
+   * @param _r ECDSA signature r
+   * @param _s ECDSA signature s
+   */
+  function permit(
+    address _owner,
+    address _spender,
+    uint256 _value,
+    uint256 _deadline,
+    uint8 _v,
+    bytes32 _r,
+    bytes32 _s
+  ) public virtual override {
+    require(block.timestamp <= _deadline, "ERC20Permit: expired deadline");
+
+    bytes32 _structHash = keccak256(
+      abi.encode(_PERMIT_TYPEHASH, _owner, _spender, _value, _useNonce(_owner), _deadline)
+    );
+
+    bytes32 _hash = _hashTypedDataV4(_structHash);
+
+    address _signer = ECDSA.recover(_hash, _v, _r, _s);
+    require(_signer == _owner, "ERC20Permit: invalid signature");
+
+    _approve(_owner, _spender, _value);
+  }
+
+  /**
+   * @dev See {IERC20Permit-nonces}.
+   */
+  function nonces(address _owner) public view virtual override returns (uint256) {
+    return _nonces[_owner].current();
+  }
+
+  /**
+   * @dev See {IERC20Permit-DOMAIN_SEPARATOR}.
+   * This is ALWAYS calculated at runtime
+   * because the token name may change
+   */
+  function DOMAIN_SEPARATOR() external view override returns (bytes32) {
+    return _domainSeparatorV4();
+  }
```

Comment:
so, in the [eip712](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/ac5931e2afbff4a66faede4c830998ee56a301c9/contracts/utils/cryptography/EIP712Upgradeable.sol#L63-L65) implementation called here it uses a stored variable of `_HASHED_NAME`. since we can update the name via the `_detailsHash` mechanism we have to call [`_buildDomainSeparator `](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/ac5931e2afbff4a66faede4c830998ee56a301c9/contracts/utils/cryptography/EIP712Upgradeable.sol#L67-L73) directly using the `keccak256(bytes(token.name))` as the name hash, and the rest of the arguments remaining the same